src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs

```diff
@@ -10,4 +10,402 @@
+        0 → V.ValueNone
+        y → V.ValueSome (V.ValuePair (V.ValueInt (x `div` y)) (V.ValueInt (rem x y)))
+    )
+
+var ∷ (Env.Instruction m, Env.Error m) ⇒ Symbol → m Env.Expanded
+var symb = do
+  stack ← get @"stack"
+  let pushStack value =
+        case VStack.lookupType symb stack of
+          Just t → do
+            modify @"stack" (VStack.cons (VStack.var1E symb (Just value), t))
+          Nothing →
+            throw @"compilationError" (Types.NotInStack symb)
+  case VStack.lookup symb stack of
+    Nothing →
+      throw @"compilationError" (Types.NotInStack symb)
+    -- TODO ∷ reduce usage by 1
+    Just (VStack.Value (VStack.Val' value)) → do
+      pushStack (VStack.ConstE value)
+      pure (Env.Constant value)
+    Just (VStack.Value (VStack.Lam' lamPartial)) → do
+      pushStack (VStack.LamPartialE lamPartial)
+      pure (Env.Curr lamPartial)
+    Just (VStack.Position usage index)
+      | one == usage →
+        Env.Expanded <$> moveToFront index
+      | otherwise →
+        Env.Expanded <$> dupToFront index
+
+-- |
+-- Name calls inst, and then determines how best to name the form in the VStack
+name ∷ Env.Reduction m ⇒ Symbol → Types.NewTerm → m Env.Expanded
+name symb f@(form, _usage, _type') = do
+  result ← inst f
+  case form of
+    Ann.Var {} →
+      modify @"stack" (VStack.nameTop symb)
+    -- all prims shouldn't add their arguments to the vstack
+    Ann.Prim {} →
+      modify @"stack" (VStack.nameTop symb)
+  -- consVar symb result usage type'
+  pure result
+
+-- for constant we shouldn't be applying it unless it's a lambda Ι don't think!?
+primToFargs ∷ Num b ⇒ Types.NewPrim → Types.Type → (Env.Fun, b)
+primToFargs (Types.Constant _) _ = undefined
+primToFargs (Types.Inst inst) ty =
+  case inst of
+    -- Can't abstract out pattern due to bad forall resolution!
+    Instr.ADD _ → (Env.Fun (add ty), 2)
+    Instr.SUB _ → (Env.Fun (sub ty), 2)
+    Instr.MUL _ → (Env.Fun (mul ty), 2)
+    Instr.EDIV _ → (Env.Fun (ediv ty), 2)
+    Instr.OR {} → (Env.Fun (or ty), 2)
+    Instr.AND _ → (Env.Fun (and ty), 2)
+    Instr.XOR _ → (Env.Fun (xor ty), 2)
+
+appM ∷ Env.Reduction m ⇒ Types.NewTerm → [Types.NewTerm] → m Env.Expanded
+appM form@(t, _u, ty) args =
+  case t of
+    Ann.Prim p →
+      let (f, lPrim) = primToFargs p ty
+          argsL = length args
+       in case argsL `compare` lPrim of
+            EQ → Env.unFun f args
+            LT → do
+              names ← reserveNames (fromIntegral lPrim)
+              -- TODO ∷ set the usage of the arguments to 1
+              let constructed = Env.C
+                    { Env.fun = f,
+                      Env.argsLeft = names,
+                      Env.left = fromIntegral lPrim,
+                      Env.captures = Set.empty,
+                      Env.ty = ty,
+                      Env.argsApplied = []
+                    }
+              apply constructed args []
+            -- this should never happen, due to type checking??
```

Comment:
Correct, we should never overapply primitives, throwing an error is the right behaviour.